src/pages/index.js

```diff
@@ -36,36 +45,82 @@ const editProfilePopup = new PopupWithForm(
   "#edit-profile-popup",
   handleProfileSave
 );
+const editProfileImagePopup = new PopupWithForm(
+  "#edit-profile-image-popup",
+  handleProfileImageSave
+);
 const addCardPopup = new PopupWithForm("#add-card-popup", handleAddCard);
+const confirmDeletePopup = new PopupConfirm('#confirm-delete-popup', handleDeleteCard);
 
 /*═══════╕
  │ CARDS │
  ╘═══════*/
-const cardSection = new Section(
-  { items: initialCards, renderer: handleCreateCard },
-  ".cards__list"
-);
+let cardSection;
+let targetCard;
 
 /*════════════════╕
  │ EVENT HANDLERS │
  ╘════════════════*/
 const imagePopup = new PopupWithImage("#image-popup");
+const deleteButton = document.querySelector("#confirm-delete-button");
 
 function handleAddCard({ title, url }) {
-  const card = {
-    name: title,
-    link: url,
-  };
-  cardSection.addItem(handleCreateCard(card));
+  api.addCard({name: title, link: url})
+  .then(cardObj => {
+    cardSection.addItem(cardObj);
+    addCardPopup.close();
+  })
+  .catch(err => console.error(err));
 }
 function handleCreateCard(card) {
-  return new Card(card, "#card-template", handleImageClick).getElement();
+  return new Card(card, "#card-template", handleImageClick, handleDeleteClick, handleLikeClick).getElement();
+}
+function handleDeleteCard() {
+  confirmDeletePopup.deleting(true);
+  api.deleteCard(targetCard)
+    .then(() => api.getPageInfo()
+      .then(([cardArray]) => {
+        console.log(cardArray);
+        cardSection.setItems(cardArray);
+        cardSection.renderItems();
+        confirmDeletePopup.close();
+      })
+    )
+    .catch(err => console.log(err));
+}
+function handleDeleteClick(card) {
+  targetCard = card.getData();
+  confirmDeletePopup.deleting(false);
+  confirmDeletePopup.open();
 }
 function handleImageClick(card) {
   imagePopup.open(card.getData());
 }
+function handleLikeClick(card) {
+  targetCard = card.getData();
+  api.toggleLike(targetCard)
+  .then(cardSection.renderItems())
+  .catch(err => console.log(err));
+}
+function handleProfileImageSave({url}) {
+  console.log(url);
+  editProfileImagePopup.saving(true);
+  profileImage.src = url;
+  api.setUserPicture(url)
+    .then((res) => {
+      console.log(res);
+      editProfileImagePopup.close();
+    })
+    .catch(err => console.error(err));
+}
 function handleProfileSave({ title, description }) {
-  user.setUserInfo(title, description);
+  editProfilePopup.saving(true);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```